src/create-client.ts

```diff
@@ -31,7 +31,11 @@ interface RedirectOptions {
   type: "sign-in" | "sign-up";
 }
 
-type State = "INITIAL" | "AUTHENTICATING" | "AUTHENTICATED" | "ERROR";
+type State =
```

Comment:
i think i stole this idea from @mthadley. I'm going to use this state object to share the authentication request in a subsequent PR. (but possibly all the memoryStorage stuff gets moved into here too in the future).

Reply:
Yeah, I did some spikes that moved more of the related data into this `State` union:

```ts
type State =
  | { tag: "INITIAL" }
  | { tag: "AUTHENTICATING"; oldAccessToken: string }
  | { tag: "AUTHENTICATED"; accessToken: string }
  | { tag: "ERROR"; retries: number }
  | { tag: "DISPOSED" };
```

I also explored abstracting the state and its changes into a class in order to facilitate some looser coupling between other components, but unclear whether it was getting overly complciated:

```ts
export class ClientState extends EventTarget {
  #state: State = { tag: "INITIAL" };

  readonly #emitter = new EventTarget();

  set current(state: State) {
    this.#state = state;
    this.#emitter.dispatchEvent(new Event("change"));
  }

  get current(): State {
    return this.#state;
  }

  toBe<T extends State>(tag: T["tag"]): Promise<T> {
    if (this.#state.tag === tag) {
      return Promise.resolve(this.#state as T);
    }

    return new Promise((resolve) => {
      const controller = new AbortController();

      this.#emitter.addEventListener(
        "change",
        () => {
          if (this.#state.tag === tag) {
            resolve(this.#state as T);
            controller.abort();
          }
        },
        { signal: controller.signal },
      );
    });
  }
}
```

Reply:
Either way, the changes here look like a good direction to me!